### DISCUSSION TITLE
Is it possible to control the pan on the Wyze cam in the Apple Home

### DISCUSSION THREAD

**Discussion by bellario89** [mrlt8/docker-wyze-bridge#412] [Q&A]
Just got a pancam and the stream works flawlessly. Is there a way to control the pan on a camera?

**Comment by mrlt8** [comment]
Should be possible, but we'd need a way to listen for commands from the user. 

Maybe MQTT or a rest api? Onvif looks ideal for cameras, but not sure if there's a server library we could use.

Open to suggestions or ideas.

**Reply by zimmra** [reply]
>Onvif looks ideal for cameras, but not sure if there's a server library we could use.
> 
> Open to suggestions or ideas.

[VISCA ](https://www.sony.net/Products/CameraSystem/CA/BRC_X1000_BRC_H800/Technical_Document/C456100121.pdf)is probably the most widely used PTZ camera control protocol, at least from everything that I see in both the CCTV and Unified-Communication markets

[Here](https://github.com/BreeeZe/rpos/blob/master/lib/PTZDriver.ts) is an example of someone that is running an ONVIF server and translating the received ONVIF-PTZ-commands into various PTZ protocols (including VISCA)

Personally, I would be happy with anything and probably be able to utilize anything one way or the other, whether its VISCA, ONVIF, MQTT, or either of the 3 'translating' to the other in the background

>we'd need a way to listen for commands from the user.

@mrlt8, do you know what the Pan Cams are expecting (in terms of protocol or command?)

**Reply by mrlt8** [reply]
I believe we should be able to control them over the tutk protocol locally similar to how we can send [commands](https://github.com/mrlt8/docker-wyze-bridge#camera-control) to turn on/off the night vision, etc.:
```
K11000_setRotaryByDegree
K11002_setRotaryByAction
K11004_resetRotatePosition
...
```

**Reply by zimmra** [reply]
> I believe we should be able to control them over the tutk protocol locally similar to how we can send [commands](https://github.com/mrlt8/docker-wyze-bridge#camera-control) to turn on/off the night vision, etc.:
> 
> ```
> K11000_setRotaryByDegree
> K11002_setRotaryByAction
> K11004_resetRotatePosition
> ...
> ```

Is there a way to test these? 

I see them referenced in ./app/wyzecam/tutk/device_config.json but not in ./app/wyzecam/tutk/tutk_protocol.py

**Reply by mrlt8** [reply]
Not yet. I still need to port them over from the app.

**Reply by AceSG1** [reply]
Is there a timeline for this? and will it be available for API?

**Reply by zimmra** [reply]
> Not yet. I still need to port them over from the app.

Here's an excerpt that I decoded from the APK

```java
    public static byte[] K11000_setRotaryByDegree(int i, int i2, int i3) {
        byte[] bArr = new byte[5];
        ByteOperator.byteArrayCopy(bArr, 0, ByteOperator.intToByteArray2(i), 0, 1);
        ByteOperator.byteArrayCopy(bArr, 2, ByteOperator.intToByteArray2(i2), 0, 1);
        bArr[4] = (byte) i3;
        return CamProtocolUtils.create(11000, 5, bArr);
    }

    public static byte[] K11002_setRotaryByAction(int i, int i2, int i3) {
        return CamProtocolUtils.create(11002, 3, new byte[]{(byte) i, (byte) i2, (byte) i3});
    }

    public static byte[] K11004_resetRotatePosition(int i) {
        return CamProtocolUtils.create(11004, 1, new byte[]{(byte) i});
    }
```

from

```smali
.method public static K11000_setRotaryByDegree(III)[B
    .locals 4

    const/4 v0, 0x5

    new-array v1, v0, [B

    .line 1
    invoke-static {p0}, Lcom/HLApi/utils/ByteOperator;->intToByteArray2(I)[B

    move-result-object p0

    const/4 v2, 0x0

    const/4 v3, 0x1

    invoke-static {v1, v2, p0, v2, v3}, Lcom/HLApi/utils/ByteOperator;->byteArrayCopy([BI[BII)Z

    .line 2
    invoke-static {p1}, Lcom/HLApi/utils/ByteOperator;->intToByteArray2(I)[B

    move-result-object p0

    const/4 p1, 0x2

    invoke-static {v1, p1, p0, v2, v3}, Lcom/HLApi/utils/ByteOperator;->byteArrayCopy([BI[BII)Z

    int-to-byte p0, p2

    const/4 p1, 0x4

    aput-byte p0, v1, p1

    const/16 p0, 0x2af8

    .line 3
    invoke-static {p0, v0, v1}, Lcom/HLApi/CameraAPI/protocol/CamProtocolUtils;->create(II[B)[B

    move-result-object p0

    return-object p0
.end method

.method public static K11002_setRotaryByAction(III)[B
    .locals 3

    const/4 v0, 0x3

    new-array v1, v0, [B

    int-to-byte p0, p0

    const/4 v2, 0x0

    aput-byte p0, v1, v2

    int-to-byte p0, p1

    const/4 p1, 0x1

    aput-byte p0, v1, p1

    int-to-byte p0, p2

    const/4 p1, 0x2

    aput-byte p0, v1, p1

    const/16 p0, 0x2afa

    invoke-static {p0, v0, v1}, Lcom/HLApi/CameraAPI/protocol/CamProtocolUtils;->create(II[B)[B

    move-result-object p0

    return-object p0
.end method

.method public static K11004_resetRotatePosition(I)[B
    .locals 3

    const/4 v0, 0x1

    new-array v1, v0, [B

    int-to-byte p0, p0

    const/4 v2, 0x0

    aput-byte p0, v1, v2

    const/16 p0, 0x2afc

    invoke-static {p0, v0, v1}, Lcom/HLApi/CameraAPI/protocol/CamProtocolUtils;->create(II[B)[B

    move-result-object p0

    return-object p0
.end method
```

I started to mess with adding it in to tutk_protocol.py but am not versed in java nor accustomed to dealing with byte arrays so will be a research project for me to get any further than I have

If there's anything I can help out with to progress this please let me know!

**Reply by mrlt8** [reply]
Don't have a pan cam so I can't test it out myself, but I pushed some changes to the dev branch:

https://github.com/mrlt8/docker-wyze-bridge/compare/main...dev

**Reply by zimmra** [reply]
> Don't have a pan cam so I can't test it out myself, but I pushed some changes to the dev branch:
> 
> [main...dev](https://github.com/mrlt8/docker-wyze-bridge/compare/main...dev)

Awesome, thank you! Some issues:

The commands are sent and received fine, but there seems to be a missing duration parameter (or something similar) to control the distance traveled. 

Currently, sending rotate_{direction} causes the camera to travel that direction until it can't. 

As an example: 
- rotate_left sends camera to -180, {current y}
- rotate_right sends camera to 180, {current y}
- rotate_up sends camera to {current x}, 90
- rotate_down sends camera to {current x}, -90

*Above assumes the following params:*

- Center being 0,0
- Left/right 360-degree axis
- Up/down 180-degree axis

```
2023/02/01 11:22:50 [WyzeBridge][INFO][Driveway Cam] [CONTROL] Request: rotate_left via MQTT!
2023/02/01 11:22:50 [wyzecam.tutk.tutk_ioctl_mux][DEBUG][Driveway Cam] Now listening on channel id 0
2023/02/01 11:22:50 [wyzecam.tutk.tutk_ioctl_mux][DEBUG][Driveway Cam] SEND <K11002SetRotaryByAction code=11002 resp_code=11003> <TutkWyzeProtocolHeader prefix=b'HL' protocol=1 code=11002 txt_len=3> b'\x01\x00\x05'
2023/02/01 11:22:50 [wyzecam.tutk.tutk_ioctl_mux][DEBUG][Driveway Cam] RECV <TutkWyzeProtocolHeader prefix=b'HL' protocol=25 code=11003 txt_len=1>: b'\x01'
2023/02/01 11:22:51 [wyzecam.tutk.tutk_ioctl_mux][DEBUG][Driveway Cam] No longer listening on channel id 0
2023/02/01 11:22:51 [WyzeBridge][INFO][Driveway Cam] [CONTROL] Response: {'cmd': 'rotate_left', 'status': 'success', 'response': 1}
```

**Reply by mrlt8** [reply]
Thanks for the feedback! I [added](https://github.com/mrlt8/docker-wyze-bridge/commit/7744dfef560ff549886fa21aaf2bac920856fa06) `rotate_stop` which should send `0,0` to `K11002SetRotaryByAction` and hopefully stop the rotation.

We may be able to use `K11000SetRotaryByDegree` to manually set the position, but I'm not sure what values to send so I'll have to dig into the app to see what they're doing.

**Reply by zimmra** [reply]
Unfortunately no dice, throws the following
```
rotate_stop
{
  "cmd": "rotate_stop",
  "status": "error",
  "response": "AV_ER_SENDIOCTRL_ALREADY_CALLED"
}
```

From the digging I've done it is waiting for [`avSendIOCtrlExit`](https://www.sunipcam.com/sdk/UserManual/APIRefGuide/_a_v_a_p_is_8h.html#afe09d74c629c7c86ab0c8a3564677dc4) or perhaps the [disconnect function ](https://github.com/kroo/wyzecam/blob/2313de655cbde6aebb476acb33e62a359a39128e/wyzecam/iotc.py#L725-L732)

**Reply by mrlt8** [reply]
I added some new [set_rotary](https://github.com/mrlt8/docker-wyze-bridge/commit/d79be49f639d77e00a5538deccdbcf49f6e45e6f) commands for testing. I believe the app may doing. `(+/-40,0)` for left/right and `(0,+/-25)` for up down..

**Reply by zimmra** [reply]
> I added some new [set_rotary](https://github.com/mrlt8/docker-wyze-bridge/commit/d79be49f639d77e00a5538deccdbcf49f6e45e6f) commands for testing. I believe the app may doing. `(+/-40,0)` for left/right and `(0,+/-25)` for up down..

More progress!

The following work (commands that send view to the right)
- set_rotary_10
	-  Not much movement, useless
- set_rotary_40
	-	Not much movement, still mostly useless except for fine nudges
- set_rotary_90
	- Good 'small' move to the right
- set_rotary_180
	- Good 'standard' move to the right

The camera does move, but all of these do throw an error in the response though

```
set_rotary_90
{
  "cmd": "set_rotary_90",
  "status": "error",
  "response": "timeout"
}
```

The set_rotary_m## commands do not work at all
```
set_rotary_m90
{
  "cmd": "set_rotary_m90",
  "status": "error",
  "response": "can't convert negative int to unsigned"
}
```

From logs:
```log
2023/02/06 14:11:17 [WyzeBridge][INFO][Driveway Cam] [CONTROL] Request: set_rotary_m90 via MQTT!
2023/02/06 14:11:17 [wyzecam.tutk.tutk_ioctl_mux][DEBUG][Driveway Cam] Now listening on channel id 0
2023/02/06 14:11:17 [wyzecam.tutk.tutk_ioctl_mux][DEBUG][Driveway Cam] RECV <TutkWyzeProtocolHeader prefix=b'HL' protocol=25 code=10059 txt_len=1>: b'\x01'
2023/02/06 14:11:17 [wyzecam.tutk.tutk_ioctl_mux][DEBUG][Driveway Cam] No longer listening on channel id 0
2023/02/06 14:11:17 [WyzeBridge][WARNING][Driveway Cam] [CONTROL] can't convert negative int to unsigned
```

**Reply by mrlt8** [reply]
Thanks for the info! Pushed some changes for the unsigned error:
https://github.com/mrlt8/docker-wyze-bridge/commit/cf9351cfb66e7829b3ef7237daff51e81a634598

You should be able to ignore the timeout error - I believe it's because the camera doesn't send back any data for the `setRotaryByDegree` command..?

Left/right should move by 180 and up/down by 90. Let me know if this needs further adjustments.

**Reply by zimmra** [reply]
> Thanks for the info! Pushed some changes for the unsigned error: [cf9351c](https://github.com/mrlt8/docker-wyze-bridge/commit/cf9351cfb66e7829b3ef7237daff51e81a634598)

Doesn't have the intended effect
With code as-is, when trying to do left/right I get the below

```
{
  "cmd": "set_rotary_left",
  "status": "error",
  "response": "int too big to convert"
}
```

The up command 'did something', but I noticed camera was moving right when issuing the up command

Looking at the code the vertical degree is in the wrong spot

To get an idea what was happening with each command (and to get the number in the -128 through 127 byte range for the command to be succsessful) I changed your code to the following:

```
K11000SetRotaryRight = partial(K11000SetRotaryByDegree, 90, 0)
K11000SetRotaryLeft = partial(K11000SetRotaryByDegree, -90, 0)
K11000SetRotaryUp = partial(K11000SetRotaryByDegree, 0, 90)
K11000SetRotaryDown = partial(K11000SetRotaryByDegree, 0, -90)
```

All of these commands result in the camera moving in a 'positive' (UP or RIGHT) direction
-90 is (seemingly) a bigger 'positive' move than 90, so set_rotary_left moves further to the right than set_rotary_right does

**Reply by mrlt8** [reply]
@zimmra, I pushed a couple of changes to the dev branch which should fix the byte issue. 

Response should show null/None if the camera doesn't return any data.

Also added an alternate `pack_rotary_up` for testing that should give the same result but will pack the values instead of doing all the slicing on the byte array.

**Reply by zimmra** [reply]
Getting very close!

I have the same results with both the pack_rotary commands and the set_rotary commands

Up and Down both work as intended!

What doesn't work is left/right. In MQTT the responses all appear to be the same, but the camera does not move left or right

**Reply by mrlt8** [reply]
Hmm, I wonder if 180 was too high for left/right? The app seems to use 40 and 90 in some places so I reduced it to 90 and added a `set_rotary_right_40` and `set_rotary_up_40` for testing.
https://github.com/mrlt8/docker-wyze-bridge/commit/fd644ffac4efee5d880f3b68cf77fff7a8c73163

**Reply by zimmra** [reply]
Up/Down/Left/Right area all working now!

I think the issue with my left/right is the way I have the camera mounted, it isn't level so I think it is getting stuck in the middle.

I'll try messing around with the speed value to see if that helps power through that a bit, as the Wyze app doesn't seem to have any problem moving through it.

**Comment by bellario89** [comment]
This is amazing, is there a way to move to a specific position? Or is it best to use the set rotary command several times

**Reply by zimmra** [reply]
In theory it's possible to modify the code to allow for an input variable, but as it currently is, it is basic up/down/left/right movement

**Reply by sibero80** [reply]
Amazing step forward! It would be a dream if we could set a specific position, for example point the camera towards a door when a contact sensor is triggered.

**Reply by mrlt8** [reply]
something with `K11012_setCruisePoints` and/or `K11016_setCruise`?

**Reply by bellario89** [reply]
Yeah. Just add another use case: we are looking for a privacy mode of sorts that move the camera out of view and then back again when we're not there

**Reply by mrlt8** [reply]
Added `K11018SetPTZPosition` in https://github.com/mrlt8/docker-wyze-bridge/commit/ddc891ee60f19703c25740c2bc37b2a550634b34.

Not really sure about the values, so just added a couple for testing.

**Reply by sibero80** [reply]
Thanks @mrlt8 for this commit! just gave it a try but does not seems to be working for me. Getting: `{"command":"set_ptz_10","response":"Unknown command","status":"error"}`

**Reply by mrlt8** [reply]
@sibero80 changes are on the dev branch, so you may need to pull the image again or rebuild the container.

**Reply by sibero80** [reply]
@mrlt8 Here's what I did:
`git clone -b dev https://github.com/mrlt8/docker-wyze-bridge`
`sudo docker-compose up`
Heres a capture of a search for the added commands in the `wyze_control.py` file.
![image](https://user-images.githubusercontent.com/28715756/219526639-25bb2f12-6a1e-4b66-8420-7426121f9a1b.png)

**Reply by mrlt8** [reply]
@sibero80 Are you still getting "Unknown command" on the dev branch?

You might need to double check your docker-compose.yml to either build the image locally:
```yaml
build:
    context: ./app
```
or just pull the pre-built dev image:
```yaml
image: mrlt8/wyze-bridge:dev
```

**Reply by sibero80** [reply]
@mrlt8 it seems I'm not skilled enough to make this work in the dev branch, although in my latest attempr I saw the version # go to the latest V2.0, but I still cannot make the cammera move with the added commands.  Will paciently wait for the stable release.  Thank you!

**Reply by bellario89** [reply]
Hi just wanted to update after testing with 2.0

Try each of the ptz commands and was not able to rotate the camera!

**Reply by bellario89** [reply]
Happy to try a fork and play around with it, where did you find the API docs to move with the other commands?

**Reply by mrlt8** [reply]
Thanks for testing @bellario89! 
Unfortunately, there is no documentation for the wyze commands and they need to be reverse engineered from the wyze app. Easiest way would probably be to decompile the android apk with something like jadx and try to figure out the commands in `/sources/com/HLApi/CameraAPI/protocol/CamCommand.java`

**Reply by bellario89** [reply]
Ah shoot that may be beyond my capacity but I'll try to take a look.
